scvi/dataset/dataset.py

```diff
@@ -3,275 +3,888 @@
+            Or boolean array used as a mask-like index.
+        """
+        nb_cells_old = self.nb_cells
+
+        # update cell-related attributes accordingly
+        for attribute_name in self.cell_attribute_names:
+            attr = getattr(self, attribute_name)
+            setattr(self, attribute_name, attr[subset_cells])
+
+        # subsample datasets which calls library_size_batch which requires already updated attributes
+        self.X = self.X[subset_cells]
+        for version_name in self.dataset_versions:
+            data = getattr(self, version_name)
+            setattr(self, version_name, data[:, subset_cells])
+
+        logging.info(
+            "Downsampled from {nb_cells} to {new_nb_cells} cells".format(
+                nb_cells=nb_cells_old, new_nb_cells=self.nb_cells
+            )
+        )
+
+    def subsample_genes(
+        self,
+        new_n_genes: Optional[int] = None,
+        new_ratio_genes: Optional[float] = None,
+        subset_genes: Optional[Union[List[int], List[bool], np.ndarray]] = None,
+    ):
+        """Wrapper around ``update_genes`` allowing for manual and automatic (based on count variance) subsampling.
+
+        The function either:
+            * Subsample `new_n_genes` genes among all genes
+            * Subsamble a proportion of `new_ratio_genes` of the genes
+            * Subsample the genes in `subset_genes`
+
+        :param subset_genes: list of indices or mask of genes to retain
+        :param new_n_genes: number of genes to retain, the highly variable genes will be kept
+        :param new_ratio_genes: proportion of genes to retain, the highly variable genes will be kept
+        """
+        if new_ratio_genes is not None:
+            if 0 < new_ratio_genes < 1:
+                new_n_genes = int(new_ratio_genes * self.nb_genes)
+            else:
+                logger.info(
+                    "Not subsampling. Expecting 0 < (new_ratio_genes={new_ratio_genes})  < 1.".format(
+                        new_ratio_genes=new_ratio_genes
+                    )
+                )
+                return
+
+        if new_n_genes is not None:
+            if new_n_genes >= self.nb_genes or new_n_genes < 1:
+                logger.info(
+                    "Not subsampling. Expecting: 1 < (new_n_genes={new_n_genes}) <= self.nb_genes".format(
+                        new_n_genes=new_n_genes
+                    )
+                )
+                return
+
             std_scaler = StandardScaler(with_mean=False)
             std_scaler.fit(self.X.astype(np.float64))
             subset_genes = np.argsort(std_scaler.var_)[::-1][:new_n_genes]
-        self.update_genes(subset_genes)
 
-    def filter_genes(self, gene_names_ref, on='gene_names'):
+        if subset_genes is None:
+            logger.info(
+                "Not subsampling. No parameter given".format(new_n_genes=new_n_genes)
+            )
+            return
+
+        self.update_genes(np.array(subset_genes))
```

Comment:
Would it be clearer to construct and return a new `GeneExpressionDataset` rather than to mutate the current `GeneExpressionDataset`?

Reply:
And then delete the `update_cells` method?

Reply:
But just an idea...feel free not to incorporate it. This PR is already really impressive.

Reply:
Actually all the methods in GeneExpressionDataset are inplace:
- It was the former behavior
- With huge dataset we cannot duplicate the data for each update

Reply:
Glad you changed that! Formerly numpy was called in a way that made a copy of the count matrix -- not very space efficient.